pkgs/build-support/trivial-builders/default.nix

```diff
@@ -127,7 +127,14 @@ rec {
 
   # See doc/build-helpers/trivial-build-helpers.chapter.md
   # or https://nixos.org/manual/nixpkgs/unstable/#trivial-builder-text-writing
-  writeText = name: text: writeTextFile { inherit name text; };
+  writeText = name: text: writeTextFile {
+    inherit name text;
+    derivationArgs = rec {
+      outputHashAlgo = "sha256";
+      outputHashMode = "flat";
+      outputHash = builtins.hashString outputHashAlgo (builtins.toString text);
```

Comment:
```suggestion
      outputHash = builtins.hashString outputHashAlgo text;
```

Reply:
This was my original approach but it breaks a test which passes a derivation as the text, and relies on implicit toString functionality somewhere to convert it to its string representation. builtins.hashString is less forgiving and requires the input to already be a string. I'll try and find it again and add a comment instead.

Reply:
Oh I see, yeah then we can't do that, but in the future we should consider deprecating non-string arguments to `writeText`, because this behavior can easily lead to bugs.

Reply:
The test is in https://github.com/nixos/nixpkgs/blob/f78bb370168b06e51af27ada73618b12525a415c/pkgs/build-support/trivial-builders/test/concat-test.nix#L5 -- should I just add builtins.toString there, instead?

Reply:
Since `writeText` can be used by code outside Nixpkgs, I'd prefer to still support non-strings for now, but adding `toString` there is good in any case.

If you want to go for deprecating non-string arguments:

```nix
# TODO: To fully deprecate, replace the assertion with `lib.isString` and remove the warning
assert assertMsg (lib.strings.isConvertibleWithToString text) ''
  pkgs.writeText ${lib.strings.escapeNixString name}: The second argument should be a string, but it's a ${builtins.typeOf text} instead.'';
lib.warnIf (! lib.isString text) ''
  pkgs.writeText ${lib.strings.escapeNixString name}: The second argument should be a string, but it's a ${builtins.typeOf text} instead, which is deprecated. Use `toString` to convert the value to a string first.''
writeTextFile {
  inherit name name;
  # ...
    outputHash = builtins.hashString outputHashAlgo (builtins.toString text);
}
```

Reply:
Done. I tried moving it to `writeTextFile` but that caused an infinite recursion.. 🤷‍♀️ so I've left it here.